proposals/2024/05-10_OpenID-Connect.md

```diff
@@ -0,0 +1,467 @@
+For every system accessed in every workflow, an account with access must be provisioned and stored in multiple places, 
+There are some serious negative consequences as a result:
+- labor costs and time to market are increased due to the steps needed to configure and manage the accounts
+- security risks are increased with static passwords being stored in additional systems being the system of record
+- outage risks are increased as passwords are rotated or expire
+
+OpenID Connect (OIDC) is a technology that can address all of these issues. In a CI system, short-lived OIDC tokens (implemented as JWT tokens) are created
+for each build, and each cloud provider authenticates the token with the Vela server. The OIDC token allows for fine-grained authorization with
+access to build information like repository organization, repository, author, event type, etc..
+
+
+**Please provide a summary of the new feature, redesign or refactor:**
+
+Support the OpenID Connect specification in Vela, allowing it to perform the role of OIDC Provider. [Specification implementor's guide](https://openid.net/specs/openid-connect-basic-1_0.html)
+
+The following diagram shows an example flow:
+```mermaid
+sequenceDiagram
+    participant vs as Vela Server
+    participant vw as Vela Worker
+    participant cp as Cloud Provider
+    participant r as Cloud Managed Resource
+    vs ->> vw: Start job(OIDC_TOKEN)
+    vw ->> cp: Exchange for provider access token(OIDC_TOKEN)
+    activate cp
+        cp -->> vw: Return CLOUD_PROVIDER_TOKEN
+    deactivate cp
+    vw ->> r: Configure cluster(CLOUD_PROVIDER_TOKEN)       
+    vw ->> r: Deploy cluster(CLOUD_PROVIDER_TOKEN)
+```
+
+**Please briefly answer the following questions:**
+
+1. Why is this required?
+
+The end result is that people using Vela can access resources without needing to maintain credentials, yet the information exchange is still strongly authenticated and authorized.
+This increases productivity, strengthens security, and eliminates outages caused by expired or incorrect credentials.
+
+2. If this is a redesign or refactor, what issues exist in the current implementation?
+
+N/A
+
+3. Are there any other workarounds, and if so, what are the drawbacks?
+
+This is a new feature. Users would still be able to use the existing password management features.
+
+4. Are there any related issues? Please provide them below if any exist.
+
+
+## Design
+
+**Example Pipeline**
+
+High level, this would be the necessary first step to receive an ID token for any Vela build.
+
+```yaml
+version: "1"
+
+steps:
+  - name: request token
+    image: alpine:latest
+    id_request: write  # becomes a claim in token `request`
+    commands:
+      - apk add curl
+      - 'curl -H "Authorization: Bearer $VELA_ID_TOKEN_REQUEST_TOKEN" $VELA_ID_TOKEN_REQUEST_URL'
+```
+
+
+**New Token Types**
+
+### ID_TOKEN_REQUEST_TOKEN
+
+Token is signed by _VELA_SERVER_PRIVATE_KEY_ using HS256 algorithm. It has the following claims:
```

Comment:
I think converting all our JWTs to the RS256 method is probably the way to go but figured that transition was a tad out of scope for this proposal. 

But in terms of this specific new token (`VELA_ID_TOKEN_REQUEST_TOKEN`), should we implement a server-side parser of the RSA keys and have it signed by the same private key as the `ID_TOKEN`? Or add another token type to the HMAC set with plans to migrate everything at a later time?

Reply:
> converting all our JWTs to the RS256 method is probably the way to go but figured that transition was a tad out of scope for this proposal.

i agree with all of that

Reply:
i picture rsa key parsing to be handled on the client but if you want to explain this a bit more maybe im not understanding the benefits fully

Reply:
i think its fine to have the request tokens and the id tokens signed by the same key but i havent looked into it much to know if theres a "standard" for that. im sure if you ask a security pro theyll probably suggest we dont reuse keys 😅